SparkFun Qwiic Pro Micro USB-C Atmega32u4

cecton opened the pull request:
Hello @Rahix 

I just started making the sparkfun-pro-micro crate by copying and renaming everything from the arduino-leonardo one (26698a0a191).

> @cecton, for the pins, just reference the [schematic for your board](https://cdn.sparkfun.com/assets/4/4/f/2/a/Qwiic_Pro_Micro_V2_0_USB_C_Schematic.pdf). It shows quite well which MCU pin is named what on the PCB.

It honestly took me some time to see it on the schematics :sweat_smile: I used the right most diagram.

It doesn't really show D11 to D13 but I supposed this is assumed because of the order? So I did the same with A4-A5 (Please double check that b66921e87)

Closes #52

Rahix commented on boards/sparkfun-pro-micro/Cargo.toml:
```diff
@@ -0,0 +1,19 @@
+[package]
+name = "sparkfun-pro-micro"
+version = "0.1.0"
+authors = ["Rahix <rahix@rahix.de>"]
```
That should be you :)

cecton replied:
Yeay \o/ Thanks!!

Rahix reviewed (commented):
Looks good, I have two more things:

1. Please add a section in README, like there is for all the other boards.
2. Please add the board to `.travis.yml` so CI test-builds all examples.

cecton commented:
Ok I have changed what you asked but there are 2 things I would like to discuss with you:

1. I noticed in the examples of the leonardo that the port F is actually used in the example leonardo-adc. But then why it isn't in the pins file instead? Because if it were, the leonardo and the promicro would be the exact same board.
2. The example i2cdetect actually crashes on the pro micro. I'm not certain and I'm not sure how I can debug it. I think it crashes because when I run it with the screen connected on qwiic, I get this:

```
Write direction test:
-    0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:       -- -- -- -- -- -- -- -- -- -- -- -- -- --
10: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
20: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
30: -- -- -- -- -- -- -- -- -- -- -- -- 3c -- -- --
40: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
50: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
60: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
70: -- -- -- -- -- -- -- --

Read direction test:
-    0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:       -- -- -- -- -- -- -- -- -- -- -- -- -- --
10: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
20: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
30: -- -- -- -- -- -- -- -- -- -- -- -- 3c
```

While when the screen is not connected I get this:

```
Write direction test:
-    0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:       -- -- -- -- -- -- -- -- -- -- -- -- -- --
10: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
20: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
30: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
40: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
50: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
60: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
70: -- -- -- -- -- -- -- --

Read direction test:
-    0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:       -- -- -- -- -- -- -- -- -- -- -- -- -- --
10: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
20: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
30: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
40: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
50: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
60: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
70: -- -- -- -- -- -- -- --
```

I also tried to light a led right after the read and it doesn't work. So I guess it's crashing somehow.

cecton commented:
I couldn't reproduce the error on the CI so I did an rustup update and now nothing compiles XD

Rahix commented:
@cecton, you'll need to rebase your branch on latest master, then you should be able to reproduce the error.  From the CI log, I think you're missing the prescaler value for PWM.  Should be easy to fix, look at the current version of the example in the leonardo crate.

cecton commented:
I thought I did update but I updated on my master ...

cecton commented:
It's ready

cecton commented:
@Rahix hey o/ if you have some time, could you answer my questions? I think there are improvements to do, I can open separate tickets since 2. is probably a bug but I'm not sure about 1.

Rahix commented:
> I noticed in the examples of the leonardo that the port F is actually used in the example leonardo-adc. But then why it isn't in the pins file instead?

Right, the real reason is that I didn't want to add it until ADC was implemented ... Which it now is, so this should definitely be fixed!

> Because if it were, the leonardo and the promicro would be the exact same board.

Hmm, maybe we should then go the route I dedided for with "Arduino Nano" (described in [issue #64](https://github.com/Rahix/avr-hal/issues/64#issuecomment-683676111) and implemented in [pr #69](https://github.com/Rahix/avr-hal/pull/69))? I.e. add a feature-flag for the promicro to `arduino-leonardo` and gate any promicro specific functionality behind that (if there is any).

> The example i2cdetect actually crashes on the pro micro. I'm not certain and I'm not sure how I can debug it. I think it crashes because when I run it with the screen connected on qwiic, [...]

This is most likely not a bug in the i2cdetect code but caused by an unfortunate design flaw in I2C.  From the Linux `i2cdetect` man-page:

```text
WARNING
        This program can confuse your I2C bus, cause data loss and worse! 
```

And an [excerpt from the source](https://git.kernel.org/pub/scm/utils/i2c-tools/i2c-tools.git/tree/tools/i2cdetect.c#n100):

```c
/* Probe this address */
switch (cmd) {
default: /* MODE_QUICK */
  /* This is known to corrupt the Atmel AT24RF08
     EEPROM */
  res = i2c_smbus_write_quick(file,
        I2C_SMBUS_WRITE);
  break;
case MODE_READ:
  /* This is known to lock SMBus on various
     write-only chips (mainly clock chips) */
  res = i2c_smbus_read_byte(file);
  break;
}
```

Essentially, the problem is that I2C does not have a dedicated operation for probing addresses.  Instead, `i2cdetect` implementations attempt to read from each address / write to each address and whenever a slave device responds with `ACK`, it knows something is connected at that address.  However, some devices can't deal with random reads (e.g. some write-only devices) or writes (e.g. because the device interprets the written value) and enter a weird state where they lock up the bus.  Most likely, that's why it's hanging for you ...

cecton commented:
> Right, the real reason is that I didn't want to add it until ADC was implemented ... Which it now is, so this should definitely be fixed!

Ok I will do it! :)

> Hmm, maybe we should then go the route I dedided for with "Arduino Nano" (described in issue #64 and implemented in pr #69)? I.e. add a feature-flag for the promicro to arduino-leonardo and gate any promicro specific functionality behind that (if there is any). 

I guess I can try to do that too. The code is exactly the same for leonardo and this promicro.

> This is most likely not a bug in the i2cdetect code but caused by an unfortunate design flaw in I2C. From the Linux `i2cdetect` man-page:

Linux? Sorry I meant the i2cdetect function of avr-hal: https://rahix.github.io/avr-hal/atmega32u4_hal/i2c/struct.I2c.html#method.i2cdetect

Rahix commented:
> Linux? Sorry I meant the i2cdetect function of avr-hal

Sure, I just referred to Linux as they do the exact same thing and it's better documented over there ;)

cecton commented:
> > Hmm, maybe we should then go the route I dedided for with "Arduino Nano" (described in issue #64 and implemented in pr #69)? I.e. add a feature-flag for the promicro to arduino-leonardo and gate any promicro specific functionality behind that (if there is any).
> 
> I guess I can try to do that too. The code is exactly the same for leonardo and this promicro.

I checked the code and I see the point. I said that the leonardo and promicro are exactly the same but I lied. The promicro has only 2 leds. Does that mean I get the leonaro a feature flag of the pro micro?

I also realized that I might have added too many pins. i.e. there are no pin A4 and A5 that I can see on the board! I probably made a mistake.

```rust
        /// `A0`
        pub a0: portf::pf7::PF7,
        /// `A1`
        pub a1: portf::pf6::PF6,
        /// `A2`
        pub a2: portf::pf5::PF5,
        /// `A3`
        pub a3: portf::pf4::PF4,
        /// `A4`
        pub a4: portf::pf1::PF1,
        /// `A5`
        pub a5: portf::pf0::PF0,
```

(Also not existing: D11-D13 but I added them in the code)

All of this makes me realize that promicro is really a subset of leonardo. If I make promicro a feature flag I will have put only `#[cfg(not(feature = "promicro"))]` and if I make leonardo a feature flag I will have to put only `#[cfg(feature = "leonardo")]` which is simpler (especially if we add more boards. But on the other hand, leonardo was first but, again, everything is in `0.` version so you can afford to do it easily.

Or maybe we can keep all the duplicated code for now. Using the build.rs it is probably possible to copy the code on fly during the compilation and have separate crates.

Rahix commented:
I would argue for the `#[cfg(not(feature = "promicro"))]` method.  This way the crate by default is everything leonardo and each feature flag just denotes the deviations.  Of course, for more complex scenarios one would need to carefully consider which board to use as the "base" but here I think the Leonardo is a safe bet.

Rahix commented:
> Or maybe we can keep all the duplicated code for now. Using the build.rs it is probably possible to copy the code on fly during the compilation and have separate crates.

Well worst case we shouldn't use build.rs but go with `include!()` but I would very much like to avoid going that route ... If I understand you correctly, the two boards are similar enough that it makes sense to put them in a single crate, similar to _Arduino Uno_ and _Arduino Nano_.

cecton commented:
I think they're similar only because the base --here the microchip-- is the same. If I understand things correctly, some things are connected, some are not, but the base should probably be the microschip and not a board.

I think the feature was okay for arduino uno/nano but it feels wrong to "hide" sparkfun's product behind a crate called "arduino". This only increases the confusion many people (including me) have about what is exactly an Arduino. For that reason I do prefer promicro getting its own crate.
